src/app/main/tor_threads.c

```diff
@@ -0,0 +1,43 @@
+/* Copyright (c) 2003-2004, Roger Dingledine
+ * Copyright (c) 2004-2006, Roger Dingledine, Nick Mathewson.
+ * Copyright (c) 2007-2019, The Tor Project, Inc. */
+/* See LICENSE for licensing information */
+
+#include "app/main/tor_threads.h"
+
+#include "lib/thread/threads.h"
+#include "app/main/subsysmgr.h"
+
+/** Wraps a void (*)(void*) function and its argument so we can
+ * invoke them with proper subsystem thread management.
+ */
+typedef struct tor_thread_helper_info_t {
+  void (*func)(void *);
+  void *data;
+} tor_thread_helper_info_t;
+
+static void tor_thread_helper(void *_info) ATTR_NORETURN;
+
+static void
+tor_thread_helper(void *_info)
+{
+  tor_thread_helper_info_t *info = _info;
+  void (*func)(void*) = info->func;
+  void *data = info->data;
+  tor_free(_info);
+
+  subsystems_thread_init();
+  func(data);
```

Comment:
Do we need to document somewhere that "func" must exit by returning, and that it must not call `spawn_exit()`?

Alternatively, should we fix `spawn_exit()` so that it is guaranteed to call subsystems_thread_cleanup?